### ISSUE TITLE
Need a mechanism to provide a custom `collate_fn` to PyTorch's DataLoader

### ISSUE THREAD

**Issue by drewoldag** [lincc-frameworks/fibad#491]
**Feature request**
We've been using the default collate_fn provided by PyTorch in DataLoader this entire time, but we've just run into the case with the AppleCider incubator where that is no longer sufficient. 

We should provide two things:
1. A default that just uses PyTorch's default collate_fn
2. A Hyrax collate_fn that will do automatic padding and collation given that we know that our dataset classes will always return dictionaries
3. A mechanism to allow users to provide their own collate_fn. This will likely be very similar to the various other locations for extensibility that we currently support (model, dataset, criterion and optimizer)
